tempesta_fw/http_parser.c

```diff
@@ -1812,41 +1812,144 @@ __req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len)
-		TRY_STR("*/*", Req_I_Accept, Req_I_AcceptHtml);
+		TRY_STR("text", Req_I_Accept, Req_I_AfterText);
+		TRY_STR_INIT();
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_StarSlash);
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Type);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_AfterText) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_AfterTextSlash);
+
+		__FSM_I_MOVE(Req_I_Type);
+	}
+
+	__FSM_STATE(Req_I_AfterTextSlash) {
+		TRY_STR("html", Req_I_AfterTextSlash, Req_I_AcceptHtml);
 		TRY_STR_INIT();
-		__FSM_I_JMP(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_StarSlash) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_StarSlashStar);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_StarSlashStar) {
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_AcceptHtml2);
+		return CSTR_NEQ;
 	}
 
 	__FSM_STATE(Req_I_AcceptHtml) {
 		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
 			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
 			__FSM_I_JMP(I_EoT);
 		}
-		__FSM_I_MOVE(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_AcceptHtml2) {
+		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
+			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
+			__FSM_I_JMP(I_EoT);
+		}
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Type) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Type);
+		c = *(p + __fsm_sz);
+		if (c == '/')
+			__FSM_I_MOVE_n(Req_I_Slash, __fsm_sz + 1);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Slash) {
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Subtype);
+		if (c == '*')
+			__FSM_I_MOVE(I_EoT);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Subtype) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Subtype);
+		c = *(p + __fsm_sz);
+		if (IS_WS(c) || c == ',' || c == ';')
+			__FSM_I_MOVE_n(I_EoT, __fsm_sz);
+		if (IS_CRLF(c)) {
+			return __data_off(p + __fsm_sz);
+		}
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_QValue) {
+		if (('0' <= c && c <= '9') || c == '.')
+			__FSM_I_MOVE(Req_I_QValue);
+		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c))
+			__FSM_I_JMP(I_EoT);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_WSAcceptOther) {
+		if (IS_WS(c))
+			__FSM_I_MOVE(Req_I_WSAcceptOther);
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_AcceptOther);
+		return CSTR_NEQ;
 	}
 
 	__FSM_STATE(Req_I_AcceptOther) {
-		__FSM_I_MATCH_MOVE(uri, Req_I_AcceptOther);
+		TRY_STR("q=", Req_I_AcceptOther, Req_I_QValue);
+		TRY_STR_INIT();
+		__FSM_I_MATCH_MOVE(token, Req_I_AcceptOther);
 		c = *(p + __fsm_sz);
-		if (IS_WS(c) || c == ',')
-			__FSM_I_MOVE_n(I_EoT, __fsm_sz + 1);
+		if (c == '=')
+			__FSM_I_MOVE_n(Req_I_ParamValue, __fsm_sz + 1);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_ParamValue) {
+		if (c == '\"')
+			__FSM_I_MOVE(Req_I_QuotedString);
+		__FSM_I_MATCH_MOVE(token, Req_I_ParamValue);
+		c = *(p + __fsm_sz);
+		if (IS_WS(c) || c == ',' || c == ';')
+			__FSM_I_MOVE_n(I_EoT, __fsm_sz);
 		if (IS_CRLF(c)) {
 			return __data_off(p + __fsm_sz);
 		}
 		return CSTR_NEQ;
 	}
 
+	__FSM_STATE(Req_I_QuotedString) {
+		if (c != '"')
```

Comment:
We can speed up quoted string matching by adding `__FSM_I_MATCH_MOVE(token, Req_I_QuotedString);` to previous line, or other alphabe will be more useful.  Idea is to grab all allowed symbols in batched mode, but stop right before closing quote.